src/Microsoft.AspNetCore.SignalR.Core/SerializedHubMessage.cs

```diff
@@ -35,38 +37,84 @@ public SerializedHubMessage(HubMessage message)
 
         public ReadOnlyMemory<byte> GetSerializedMessage(IHubProtocol protocol)
         {
-            if (!TryGetCached(protocol.Name, out var serialized))
+            // Double-check locking!
+            if (!TryGetCachedFast(protocol.Name, out var serialized))
             {
-                if (Message == null)
+                lock (_lock)
                 {
-                    throw new InvalidOperationException(
-                        "This message was received from another server that did not have the requested protocol available.");
-                }
+                    if (!TryGetCached(protocol.Name, out serialized))
+                    {
+                        if (Message == null)
+                        {
+                            throw new InvalidOperationException(
+                                "This message was received from another server that did not have the requested protocol available.");
+                        }
 
-                serialized = protocol.GetMessageBytes(Message);
-                SetCache(protocol.Name, serialized);
+                        serialized = protocol.GetMessageBytes(Message);
+                        SetCache(protocol.Name, serialized);
+                    }
+                }
             }
 
             return serialized;
         }
 
+        // Used for unit testing.
+        internal IEnumerable<SerializedMessage> GetAllSerializations()
```

Comment:
At first I was like wtf then I saw the comment 😄

Reply:
i got you fam https://github.com/aspnet/SignalR/pull/2079#discussion_r182593089

Reply:
Also, we really should load up the list in the lock (even if this is only used in unit tests, let's do it right). Changing it to load up a list in the lock and then return that.